### DISCUSSION TITLE
Alt-DSS python - Batch for Buses

### DISCUSSION THREAD

**Discussion by imperanatus** [dss-extensions/dss-extensions#39] [General]
I've written this simple code to select all 3phase Buses with KV > 1VK. 

buses = altdss.Bus.Name()
busesvoltages = altdss.Bus.kVBase()
busesnumnodes = altdss.Bus.NumNodes()

for bus in range(len(buses)):
    if buses[bus] != 'sourcebus' and busesvoltages[bus] > 1 and busesnumnodes[bus] == 3:
        mv_buses = np.append(mv_buses, buses[bus])

However, I've seen in the documentation that we can do batch features with conditionals to select Loads, but for Buses I didn't find the same feature. 

Am I doing something wrong? 

The aforementioned script works well with no problems, but, is there a batch way to do this? Thank you in advance!

**Comment by PMeira** [comment]
> Am I doing something wrong?

No, right now bus batches can't be created (directly by the user). We plan to add some support later.

Buses are not really DSS objects (you don't have an equivalent "new Bus.bus_name" like for loads and other objects), they are implicitly defined by the circuit elements. If the circuit elements that create a certain a bus are disabled, the bus ceases to exist, and that presents an extra difficulty for tracking the bus references in batches.

> The aforementioned script works well with no problems, but, is there a batch way to do this? Thank you in advance!

I don't recommend using NumPy arrays for lists of general objects or strings, a simple Python list is enough for those cases. A simpler version:

```python
busnames_hv = [
    name
    for name, kv, nnodes in zip(altdss.Bus.Name(), altdss.Bus.kVBase(), altdss.Bus.NumNodes())
    if kv > 1 and nnodes == 3
]
```

**Reply by PMeira** [reply]
Depending on what you're doing, while we don't have native dataframe yet, you could use JSON with Pandas for some operations:

```python
import json
import pandas as pd

df_bus = pd.DataFrame(json.loads(altdss.Bus.to_json()))

# This returns the bus names as a list
busnames_kv = df_bus[(df_bus['kVLN'] > 1) & (df_bus['Name'] != 'sourcebus')].Name.tolist()

# Alternatively, this adds a new HV column to the dataframe, and you can filter using it later
df_bus = df_bus.eval("HV = (kVLN > 1) & (Name != 'sourcebus')")
```

**Reply by imperanatus** [reply]
Ok! I got it. Thx, Paulo. I'll try this.
